metrics/ci_integrations/test/firestore/adapter/firestore_storage_client_adapter_test.dart

```diff
@@ -0,0 +1,137 @@
+      'with null Firestore',
+      () {
+        expect(() => FirestoreStorageClientAdapter(null), throwsArgumentError);
+      },
+    );
+
+    test(
+      '.addBuilds() should return normally if it throws GrpcError '
+      'with status code "not found"',
+      () {
+        when(documentReferenceMock.get()).thenThrow(GrpcError.notFound());
+        expect(() => mockAdapter.addBuilds(projectId, []), returnsNormally);
+      },
+    );
+
+    test(
+      '.addBuilds() should throw Exception if the exception '
+      'is different from GrpcError with status code "not found"',
+      () {
+        when(documentReferenceMock.get()).thenThrow(Exception());
+        expect(() => mockAdapter.addBuilds(projectId, []), throwsException);
+      },
+    );
+
+    test(
+      '.addBuilds() ensure that methods isnot called after throws GrpcError '
+      'with status code "not fount" ',
+      () async {
+        when(documentReferenceMock.get()).thenThrow(GrpcError.notFound());
+        await mockAdapter.addBuilds(projectId, []);
+        verifyNever(firestoreMock.collection('build'));
+      },
+    );
+
+    test('.fetchLastBuild should return null, if documents is empty', () {
+      when(collectionFetchBuildsMock.getDocuments()).thenAnswer(
+        (_) => Future.value([]),
+      );
+      expect(mockAdapter.fetchLastBuild(projectId), completion(equals(null)));
+    });
+
+    test(
+      '.fetchLastBuild should return BuildData, if documents is not empty',
+      () {
+        final documentMock = _DocumentMock(testDocumentId, buildDataTestJson);
+
+        when(collectionFetchBuildsMock.getDocuments()).thenAnswer(
+          (_) => Future.value([documentMock]),
+        );
+        expect(
+          mockAdapter.fetchLastBuild(projectId),
+          completion(
+            equals(
+              BuildDataDeserializer.fromJson(
+                buildDataTestJson,
+                testDocumentId,
+              ),
+            ),
+          ),
+        );
+      },
+    );
+  });
+}
+
+class _FirestoreMock extends Mock implements Firestore {
+  _FirestoreMock(_CollectionReferenceMockAddBuilds collectionAddBuilds,
+      _CollectionReferenceMockFetchBuild collectionFetchBuild) {
+    when(collection('projects')).thenReturn(collectionAddBuilds);
+    when(collection('build')).thenReturn(collectionFetchBuild);
+  }
+}
+
+class _CollectionReferenceMockAddBuilds extends Mock
+    implements CollectionReference {
+  _CollectionReferenceMockAddBuilds({
+    String projectId,
+    _DocumentReferenceMock documentReferenceMock,
+  }) {
+    when(document(projectId)).thenReturn(documentReferenceMock);
+  }
+}
+
+class _CollectionReferenceMockFetchBuild extends Mock
+    implements CollectionReference {
+  _CollectionReferenceMockFetchBuild({
+    String projectId,
+  }) {
+    when(where('projectId', isEqualTo: projectId)).thenReturn(this);
+    when(orderBy('startedAt', descending: true)).thenReturn(this);
+    when(limit(1)).thenReturn(this);
```

Comment:
I think this should be placed in tests since it provides unobvious dependencies between tests where this mock is used. From the [Best Practices section](https://pub.dev/packages/mockito#best-practices) of the mockito package:
> A class which extends Mock should never stub out it's own responses with when in it's constructor or anywhere else. Stubbed responses should be defined in the tests where they are used.